File: frontend/src/api/prompts.js
/**
 * API client for the generalized prompt system.
 * 
 * This module provides functions for interacting with the new unified
 * prompt management API that supports multiple entity types.
 */

import api from '../services/api';

/**
 * Entity types supported by the prompt system
 */
export const ENTITY_TYPES = {
  KNOWLEDGE_BASE: 'knowledge_base',  // For KB context prompts (assigned via model configs)
  LLM_MODEL: 'llm_model',
  AGENT: 'agent',
  WORKFLOW: 'workflow',
  TOOL: 'tool'
};

/**
 * Prompt API client
 */
export const promptAPI = {
  /**
   * Create a new prompt
   * @param {Object} promptData - Prompt creation data
   * @param {string} promptData.name - Prompt name
   * @param {string} promptData.description - Prompt description (optional)
   * @param {string} promptData.content - Prompt content/template
   * @param {string} promptData.entity_type - Entity type (knowledge_base, llm_model, etc.)
   * @param {boolean} promptData.is_active - Whether the prompt is active
   * @returns {Promise<Object>} Created prompt
   */
  async create(promptData) {
    const response = await api.post('/prompts', promptData);
    return response.data;
  },

  /**
   * Get all prompts with filtering and pagination
   * @param {Object} params - Query parameters
   * @param {string} params.entity_type - Filter by entity type (optional)
   * @param {string} params.entity_id - Filter by entity ID (optional)
   * @param {boolean} params.is_active - Filter by active status (optional)
   * @param {string} params.search - Search in name and description (optional)
   * @param {number} params.limit - Maximum number of results (default: 50)
   * @param {number} params.offset - Number of results to skip (default: 0)
   * @returns {Promise<Object>} List of prompts with metadata
   */
  async list(params = {}) {
    const response = await api.get('/prompts', { params });
    return response.data;
  },

  /**
   * Get a specific prompt by ID
   * @param {string} promptId - Prompt ID
   * @returns {Promise<Object>} Prompt data
   */
  async get(promptId) {
    const response = await api.get(`/prompts/${promptId}`);
    return response.data;
  },

  /**
   * Update an existing prompt
   * @param {string} promptId - Prompt ID
   * @param {Object} updateData - Update data
   * @param {string} updateData.name - Prompt name (optional)
   * @param {string} updateData.description - Prompt description (optional)
   * @param {string} updateData.content - Prompt content (optional)
   * @param {boolean} updateData.is_active - Active status (optional)
   * @returns {Promise<Object>} Updated prompt
   */
  async update(promptId, updateData) {
    const response = await api.put(`/prompts/${promptId}`, updateData);
    return response.data;
  },

  /**
   * Delete a prompt and all its assignments
   * @param {string} promptId - Prompt ID
   * @returns {Promise<void>}
   */
  async delete(promptId) {
    await api.delete(`/prompts/${promptId}`);
  },

  /**
   * Assign a prompt to an entity
   * @param {string} promptId - Prompt ID
   * @param {Object} assignmentData - Assignment data
   * @param {string} assignmentData.entity_id - Entity ID to assign to
   * @param {boolean} assignmentData.is_active - Whether assignment is active
   * @returns {Promise<Object>} Assignment data
   */
  async assign(promptId, assignmentData) {
    const response = await api.post(`/prompts/${promptId}/assignments`, assignmentData);
    return response.data;
  },

  /**
   * Unassign a prompt from an entity
   * @param {string} promptId - Prompt ID
   * @param {string} entityId - Entity ID to unassign from
   * @returns {Promise<void>}
   */
  async unassign(promptId, entityId) {
    await api.delete(`/prompts/${promptId}/assignments/${entityId}`);
  },

  /**
   * Get all prompts assigned to a specific entity
   * @param {string} entityId - Entity ID
   * @param {string} entityType - Entity type
   * @param {boolean} activeOnly - Return only active prompts (default: true)
   * @returns {Promise<Array>} List of assigned prompts
   */
  async getEntityPrompts(entityId, entityType, activeOnly = true) {
    const params = { entity_type: entityType, active_only: activeOnly };
    const response = await api.get(`/prompts/entities/${entityId}`, { params });
    return response.data;
  },

  /**
   * Get system-wide prompt statistics
   * @returns {Promise<Object>} System statistics
   */
  async getStats() {
    const response = await api.get('/prompts/stats');
    return response.data;
  }
};

/**
 * Knowledge Base specific prompt utilities
 * These provide convenience methods for working with knowledge base prompts
 */
/**
 * Knowledge Base specific prompt utilities
 * These provide convenience methods for working with knowledge base context prompts
 * Note: KB prompts are now assigned via model configurations, not directly to KBs
 */
export const knowledgeBasePromptAPI = {
  /**
   * Create a knowledge base context prompt
   * @param {Object} promptData - Prompt data (entity_type will be set automatically)
   * @returns {Promise<Object>} Created prompt
   */
  async create(promptData) {
    return promptAPI.create({
      ...promptData,
      entity_type: ENTITY_TYPES.KNOWLEDGE_BASE
    });
  },

  /**
   * Get all knowledge base context prompts
   * @param {Object} params - Query parameters
   * @returns {Promise<Object>} List of knowledge base prompts
   */
  async list(params = {}) {
    return promptAPI.list({
      ...params,
      entity_type: ENTITY_TYPES.KNOWLEDGE_BASE
    });
  },

  // NOTE: KB prompt assignments are now managed at the model configuration level
  // Use modelConfigAPI.assignKBPrompt() and modelConfigAPI.removeKBPrompt() instead
};

/**
 * LLM Model specific prompt utilities (for future use)
 * These provide convenience methods for working with LLM model prompts
 */
export const llmModelPromptAPI = {
  /**
   * Create an LLM model prompt
   * @param {Object} promptData - Prompt data (entity_type will be set automatically)
   * @returns {Promise<Object>} Created prompt
   */
  async create(promptData) {
    return promptAPI.create({
      ...promptData,
      entity_type: ENTITY_TYPES.LLM_MODEL
    });
  },

  /**
   * Get all LLM model prompts
   * @param {Object} params - Query parameters
   * @returns {Promise<Object>} List of LLM model prompts
   */
  async list(params = {}) {
    return promptAPI.list({
      ...params,
      entity_type: ENTITY_TYPES.LLM_MODEL
    });
  },

  /**
   * Get prompts for a specific LLM model
   * @param {string} modelId - LLM model ID
   * @param {boolean} activeOnly - Return only active prompts
   * @returns {Promise<Array>} List of assigned prompts
   */
  async getForModel(modelId, activeOnly = true) {
    return promptAPI.getEntityPrompts(modelId, ENTITY_TYPES.LLM_MODEL, activeOnly);
  },

  /**
   * Assign a prompt to an LLM model
   * @param {string} promptId - Prompt ID
   * @param {string} modelId - LLM model ID
   * @param {boolean} isActive - Whether assignment is active
   * @returns {Promise<Object>} Assignment data
   */
  async assignToModel(promptId, modelId, isActive = true) {
    return promptAPI.assign(promptId, {
      entity_id: modelId,
      is_active: isActive
    });
  },

  /**
   * Unassign a prompt from an LLM model
   * @param {string} promptId - Prompt ID
   * @param {string} modelId - LLM model ID
   * @returns {Promise<void>}
   */
  async unassignFromModel(promptId, modelId) {
    return promptAPI.unassign(promptId, modelId);
  }
};

export default promptAPI;

File: backend/src/shu/api/prompts.py
"""Generalized Prompt API endpoints for Shu.

This module provides REST API endpoints for managing prompts across
different entity types (knowledge bases, LLM models, agents, etc.).
"""

import logging

from fastapi import APIRouter, Depends, HTTPException, Path, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from ..auth.models import User
from ..auth.rbac import require_power_user
from ..core.database import get_db
from ..core.exceptions import ConflictError, ShuException, ValidationError
from ..core.response import ShuResponse
from ..schemas.prompt import (
    EntityTypeEnum,
    PromptAssignmentCreate,
    PromptAssignmentResponse,
    PromptCreate,
    PromptListResponse,
    PromptQueryParams,
    PromptResponse,
    PromptSystemStats,
    PromptUpdate,
)
from ..services.prompt_service import PromptAlreadyExistsError, PromptNotFoundError, PromptService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/prompts", tags=["prompts"])


@router.post(
    "/",
    response_model=PromptResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new prompt",
    description="Create a new prompt for a specific entity type.",
)
async def create_prompt(
    prompt_data: PromptCreate,
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    """Create a new prompt."""
    try:
        service = PromptService(db)
        prompt = await service.create_prompt(prompt_data)

        logger.info(f"Created prompt '{prompt.name}' for entity type '{prompt.entity_type}'")
        return ShuResponse.created(prompt.model_dump())

    except PromptAlreadyExistsError as e:
        logger.warning(f"Prompt creation failed: {e}")
        return ShuResponse.error(message=str(e), code="PROMPT_ALREADY_EXISTS", status_code=409)
    except Exception as e:
        logger.error(f"Unexpected error creating prompt: {e}")
        return ShuResponse.error(message="Internal server error", code="INTERNAL_SERVER_ERROR", status_code=500)


# Non-slash root alias for create (tolerate '/prompts' without trailing slash)
@router.post("", include_in_schema=False)
async def create_prompt_no_slash(
    prompt_data: PromptCreate,
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    return await create_prompt(prompt_data, current_user, db)


@router.get(
    "/",
    response_model=PromptListResponse,
    summary="List prompts",
    description="List prompts with filtering and pagination.",
)
async def list_prompts(
    entity_type: EntityTypeEnum | None = Query(None, description="Filter by entity type"),
    entity_id: str | None = Query(None, description="Filter by entity ID"),
    is_active: bool | None = Query(None, description="Filter by active status"),
    search: str | None = Query(None, description="Search in name and description"),
    limit: int = Query(50, ge=1, le=100, description="Maximum number of results"),
    offset: int = Query(0, ge=0, description="Number of results to skip"),
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    """List prompts with filtering and pagination."""
    try:
        service = PromptService(db)
        params = PromptQueryParams(
            entity_type=entity_type,
            entity_id=entity_id,
            is_active=is_active,
            search=search,
            limit=limit,
            offset=offset,
        )

        result = await service.list_prompts(params)
        logger.info(f"Listed {len(result.items)} prompts (total: {result.total})")

        return ShuResponse.success(result.model_dump())

    except Exception as e:
        logger.error(f"Unexpected error listing prompts: {e}")
        return ShuResponse.error(message="Internal server error", code="INTERNAL_SERVER_ERROR", status_code=500)


# Non-slash root alias for list
@router.get("", include_in_schema=False)
async def list_prompts_no_slash(
    entity_type: EntityTypeEnum | None = Query(None, description="Filter by entity type"),
    entity_id: str | None = Query(None, description="Filter by entity ID"),
    is_active: bool | None = Query(None, description="Filter by active status"),
    search: str | None = Query(None, description="Search in name and description"),
    limit: int = Query(50, ge=1, le=100, description="Maximum number of results"),
    offset: int = Query(0, ge=0, description="Number of results to skip"),
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    return await list_prompts(entity_type, entity_id, is_active, search, limit, offset, current_user, db)


@router.get(
    "/{prompt_id}",
    response_model=PromptResponse,
    summary="Get prompt by ID",
    description="Retrieve a specific prompt by its ID.",
)
async def get_prompt(
    prompt_id: str = Path(..., description="Prompt ID"),
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    """Get a prompt by ID."""
    try:
        service = PromptService(db)
        prompt = await service.get_prompt(prompt_id)

        if not prompt:
            return ShuResponse.error(message=f"Prompt {prompt_id} not found", code="PROMPT_NOT_FOUND", status_code=404)

        logger.info(f"Retrieved prompt '{prompt.name}' (ID: {prompt_id})")
        return ShuResponse.success(prompt.model_dump())

    except Exception as e:
        logger.error(f"Unexpected error getting prompt {prompt_id}: {e}")
        return ShuResponse.error(message="Internal server error", code="INTERNAL_SERVER_ERROR", status_code=500)


@router.put(
    "/{prompt_id}",
    response_model=PromptResponse,
    summary="Update prompt",
    description="Update an existing prompt.",
)
async def update_prompt(
    prompt_id: str = Path(..., description="Prompt ID"),
    prompt_data: PromptUpdate = ...,
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    """Update an existing prompt."""
    try:
        service = PromptService(db)

        # Check if this is a system default prompt
        existing_prompt = await service.get_prompt(prompt_id)
        if existing_prompt and existing_prompt.is_system_default:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="System default prompts cannot be modified",
            )

        prompt = await service.update_prompt(prompt_id, prompt_data)

        logger.info(f"Updated prompt '{prompt.name}' (ID: {prompt_id})")
        return ShuResponse.success(prompt.model_dump())

    except PromptNotFoundError as e:
        logger.warning(f"Prompt update failed: {e}")
        return ShuResponse.error(message=str(e), code="PROMPT_NOT_FOUND", status_code=404)
    except PromptAlreadyExistsError as e:
        logger.warning(f"Prompt update failed: {e}")
        return ShuResponse.error(message=str(e), code="PROMPT_ALREADY_EXISTS", status_code=409)
    except Exception as e:
        logger.error(f"Unexpected error updating prompt {prompt_id}: {e}")
        return ShuResponse.error(message="Internal server error", code="INTERNAL_SERVER_ERROR", status_code=500)


@router.delete(
    "/{prompt_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete prompt",
    description="Delete a prompt and all its assignments.",
)
async def delete_prompt(
    prompt_id: str = Path(..., description="Prompt ID"),
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    """Delete a prompt and all its assignments."""
    try:
        service = PromptService(db)

        # Check if this is a system default prompt
        existing_prompt = await service.get_prompt(prompt_id)
        if existing_prompt and existing_prompt.is_system_default:
            return ShuResponse.error(
                message="System default prompts cannot be deleted",
                code="SYSTEM_PROMPT_PROTECTED",
                status_code=403,
            )

        deleted = await service.delete_prompt(prompt_id)

        if not deleted:
            return ShuResponse.error(message=f"Prompt {prompt_id} not found", code="PROMPT_NOT_FOUND", status_code=404)

        logger.info(f"Deleted prompt {prompt_id}")
        return ShuResponse.no_content()

    except Exception as e:
        logger.error(f"Unexpected error deleting prompt {prompt_id}: {e}")
        return ShuResponse.error(message="Internal server error", code="INTERNAL_SERVER_ERROR", status_code=500)


@router.post(
    "/{prompt_id}/assignments",
    response_model=PromptAssignmentResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Assign prompt to entity",
    description="Assign a prompt to a specific entity.",
)
async def assign_prompt(
    prompt_id: str = Path(..., description="Prompt ID"),
    assignment_data: PromptAssignmentCreate = ...,
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    """Assign a prompt to an entity."""
    try:
        service = PromptService(db)
        assignment = await service.assign_prompt(prompt_id, assignment_data)

        logger.info(f"Assigned prompt {prompt_id} to entity {assignment_data.entity_id}")
        return ShuResponse.created(assignment.model_dump())

    except PromptNotFoundError as e:
        logger.warning(f"Prompt assignment failed: {e}")
        return ShuResponse.error(message=str(e), code="PROMPT_NOT_FOUND", status_code=404)
    except ConflictError as e:
        logger.warning(f"Prompt assignment conflict: {e}")
        return ShuResponse.error(message=str(e), code="ASSIGNMENT_ALREADY_EXISTS", status_code=409)
    except ValidationError as e:
        logger.warning(f"Prompt assignment blocked: {e}")
        return ShuResponse.error(message=str(e), code="VALIDATION_ERROR", status_code=422)
    except ShuException as e:
        logger.warning(f"Prompt assignment blocked: {e}")
        return ShuResponse.error(message=str(e), code=e.error_code, status_code=400)
    except Exception as e:
        logger.error(f"Unexpected error assigning prompt {prompt_id}: {e}")
        return ShuResponse.error(message="Internal server error", code="INTERNAL_SERVER_ERROR", status_code=500)


@router.delete(
    "/{prompt_id}/assignments/{entity_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Unassign prompt from entity",
    description="Remove a prompt assignment from a specific entity.",
)
async def unassign_prompt(
    prompt_id: str = Path(..., description="Prompt ID"),
    entity_id: str = Path(..., description="Entity ID"),
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    """Unassign a prompt from an entity."""
    try:
        service = PromptService(db)
        unassigned = await service.unassign_prompt(prompt_id, entity_id)

        if not unassigned:
            return ShuResponse.error(
                message=f"Assignment not found for prompt {prompt_id} and entity {entity_id}",
                code="ASSIGNMENT_NOT_FOUND",
                status_code=404,
            )

        logger.info(f"Unassigned prompt {prompt_id} from entity {entity_id}")
        return ShuResponse.no_content()

    except ShuException as e:
        logger.warning(f"Prompt unassignment blocked: {e}")
        return ShuResponse.error(message=str(e), code=e.error_code, status_code=400)
    except Exception as e:
        logger.error(f"Unexpected error unassigning prompt {prompt_id} from entity {entity_id}: {e}")
        return ShuResponse.error(message="Internal server error", code="INTERNAL_SERVER_ERROR", status_code=500)


@router.get(
    "/entities/{entity_id}",
    response_model=list[PromptResponse],
    summary="Get entity prompts",
    description="Get all prompts assigned to a specific entity.",
)
async def get_entity_prompts(
    entity_id: str = Path(..., description="Entity ID"),
    entity_type: EntityTypeEnum = Query(..., description="Entity type"),
    active_only: bool = Query(True, description="Return only active prompts"),
    current_user: User = Depends(require_power_user),
    db: AsyncSession = Depends(get_db),
):
    """Get all prompts assigned to a specific entity."""
    try:
        service = PromptService(db)
        prompts = await service.get_entity_prompts(entity_id, entity_type.value, active_only)

        logger.info(f"Retrieved {len(prompts)} prompts for entity {entity_id}")
        return ShuResponse.success([prompt.model_dump() for prompt in prompts])

    except Exception as e:
        logger.error(f"Unexpected error getting prompts for entity {entity_id}: {e}")
        return ShuResponse.error(message="Internal server error", code="INTERNAL_SERVER_ERROR", status_code=500)


@router.get(
    "/stats",
    response_model=PromptSystemStats,
    summary="Get system statistics",
    description="Get system-wide prompt statistics.",
)
async def get_system_stats(current_user: User = Depends(require_power_user), db: AsyncSession = Depends(get_db)):
    """Get system-wide prompt statistics."""
    try:
        service = PromptService(db)
        stats = await service.get_system_stats()

        logger.info("Retrieved prompt system statistics")
        return ShuResponse.success(stats.model_dump())

    except Exception as e:
        logger.error(f"Unexpected error getting system stats: {e}")
        return ShuResponse.error(message="Internal server error", code="INTERNAL_SERVER_ERROR", status_code=500)
